pkg/ingestor/collector/collector.go

```diff
@@ -15,4 +15,77 @@
 
 package collector
 
-type Collector interface{}
+import (
+	"context"
+
+	"github.com/guacsec/guac/pkg/ingestor/processor"
+	"github.com/sirupsen/logrus"
+)
+
+type Collector interface {
+	// Is the collector enabled
+	Enabled(ctx context.Context) bool
+	// Retrieve the artifacts from the collector
+	RetrieveArtifacts(ctx context.Context) ([]*processor.Document, error)
+	// Indicated when the collector is done
+	IsDone() bool
+	// Type of the collector
+	Type() CollectorType
+}
+
+type CollectorType string
+
+const (
+	CollectorGCS    = "GCS"
+	CollectorOCI    = "OCI"
+	CollectorPubSub = "PubSub"
+	CollectorRekor  = "Rekor"
+)
+
+var (
+	documentCollector = map[CollectorType]Collector{}
+)
+
+// RegisterDocumentCollector is used to register new collector types
+func RegisterDocumentCollector(p Collector, d CollectorType) {
+	if _, ok := documentCollector[d]; ok {
+		logrus.Warnf("the document processor is being overwritten: %s", d)
+	}
+	documentCollector[d] = p
+}
+
+// Enabled checks to see if the collector type is properly initialized
+func Enabled(ctx context.Context) bool {
+	for _, provider := range documentCollector {
+		if provider.Enabled(ctx) {
+			return true
+		}
+	}
+	return false
+}
+
+// Collect takes all the collectors and starts collecting
+// the artifacts
+func Collect(ctx context.Context) ([]*processor.Document, error) {
+	finalDocs := []*processor.Document{}
+	for _, collector := range documentCollector {
+		if !collector.Enabled(ctx) {
```

Comment:
i think we can assume that if a collector exists in the map then it should be enabled.

Reply:
changed to be concurrent